[WIP] [ConstraintGraph] Don't try to contract edge of parameter bindings with `inout` attribute

xedin opened the pull request:
Currently edge related to the parameter bindings is contracted
without properly checking if newly created equivalence class has
the same inout & l-value requirements. This patch improves the
situation by disallowing contraction of the edges related to parameter
binding constraint where left-hand side has `inout` attribute set.

Such guarantees that parameter can get `inout` type assigned when
argument gets `l-value` type.

Resolves: rdar://problem/33429010

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please clean smoke test

xedin commented:
@swift-ci please clean smoke test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please smoke test Linux platform

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please test source compatibility

rudkx commented on test/Constraints/closures.swift:
```diff
@@ -555,3 +555,26 @@ extension A_SR_5030 {
+func rdar33429010_2() {
+  let iter = I_33429010()
+  var acc: Int = 0 // expected-warning {{}}
+  let _: Int = AnySequence { iter }.rdar33429010(into: acc, { $0 + $1 })
```
This is no longer an `@lvalue` use of `$0`. Are you sure this is fixing the original issue?

xedin replied:
Yes, I can add a test which does the same thing but it would require localized `Data` class which `append` method.

xedin replied:
The problem in both cases is that overload for `rdar33429010` can't be chosen because closure parameter declaration vs. usage didn't match.

rudkx replied:
You could extend `Int` or another type with an `append` or some other function that mutates.

xedin replied:
Sounds good, let me try to do that!

xedin replied:
Added one more test which uses mutating method.

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

rudkx commented on lib/Sema/CSBindings.cpp:
```diff
@@ -25,6 +25,32 @@ ConstraintSystem::determineBestBindings() {
+  // represents declaration of the parameter and
+  // should drive the search.
+
+  SolverScope scope(*this);
```
I don't understand why you're opening a scope here.

xedin replied:
It's here to rollback changes to the graph and saved bindings at the end of this call.

rudkx commented on lib/Sema/CSBindings.cpp:
```diff
@@ -25,6 +25,32 @@ ConstraintSystem::determineBestBindings() {
+    if (rep1 != rep2)
+      mergeEquivalenceClasses(rep1, rep2, /*updateWorkList*/false);
+  }
+
```
I just tried removing all the updates to `determineBestBindings` and it looks like both the original and your new test case pass without those changes (unless I'm making a mistake).

xedin replied:
Yes both of the tests work without these changes but implicit conversion between pointers don't because in some sitatuions it would find bindings for use of the parameter before it's declaration. Observed that while building corelibs-foundation on Linux. Should probably add test for that too but forgot...

rudkx replied:
Yeah, we should have a test for that before merging.

xedin replied:
It was this `"".withCString({ UnsafeMutablePointer(mutating: $0) })`. Should I just add it as is or is there any way to reduce so it doesn't use stdlib?

rudkx replied:
I took another look at this. This seems like the wrong place to be doing something like this. It seems more appropriate for something like `matchTypes()` when called with `BindParam`.

xedin replied:
Unfortunately by the time we get to `matchTypes` in the example from my previous comment it's already too late, it does do the right thing as is, but the bindings which are supplied are incorrect. This situation was hidden by `contractEdges` but is exposed with these changes.

rudkx replied:
So the issue here is with the order in which we're attempting bindings. We shouldn't be attempting to propagate type information back from the parameter to the argument, but instead should be attempting bindings for `""` and then looking up the overloads, and using those to determine argument bindings.

This hack fixes this particular test case, but isn't a principled fix: https://gist.github.com/rudkx/b323a750686ada360dced7a5c2509c45

A more principled fix that would be more inline with how the solver works would be to add another field to `PotentialBindings` to track whether a type variable is the second type of a `BindParam` and these the last-ditch type variable to attempt to bind.

xedin committed:
[ConstraintGraph] Don't try to contract edge of parameter bindings with `inout` attribute

Currently edge related to the parameter bindings is contracted
without properly checking if newly created equivalence class has
the same inout & l-value requirements. This patch improves the
situation by disallowing contraction of the edges related to parameter
binding constraint where left-hand side has `inout` attribute set.

Such guarantees that parameter can get `inout` type assigned when
argument gets `l-value` type.

Resolves: rdar://problem/33429010

xedin commented:
@swift-ci please smoke test

xedin commented:
@moiseev Can you please take a look at the failing test on OS X, I'm curious about your thoughts? 

After I've removed contraction of bind parameters following code `_ = (0..<128).map({ String(UnicodeScalar($0)) })` fails to compile saying that `UnicodeScalar` is not unwrapped because overload picked for `..<` is `CountableRange<Int>` but currently (without this patch) it would pick `CountableRange<UInt8>` for `..<` since it matches string initializer `init(Unicode.Scalar)` and produces solution, but the problem is that the same code with range of `0..<129` is still (_currently_) going to pick `CountableRange<UInt8>` and fail at runtime. So I'm wondering if it's actually valid to modify test to `_ = (0..<128).map({ String(UnicodeScalar($0)!) })` or explicitly identify that 0..<128 is supposed to be `CountableRange<UInt8>`?...

 I think that new behavior is _more correct_ than what is happening right now, WDYT?

xedin commented:
Here is simplified example with represents failure on Linux:

```swift
class P {
  var name: String = "foo"
}

class C : P {
}

let classes = [C(), C()]
classes.map({ ($0.name, $0 as P) })
```
Given `T3 bind param T6` and `(T6, P) conv T2` relations.
The problem here is that we pick the binding for result of the closure `T2` which is `(T6, P)` before argument `T3` which is `(C)` and there is no implicit conversion from `C` to `P` defined for `BindParam` mentioned above.

xedin commented:
@rudkx Apparently if I change last expression to `[C()].map({ ($0.name, $0 as P) })` (instead of saving array into temporary variable), expression doesn't compile on the latest master because we pick bindings in invalid order even with contraction :(

xedin commented:
I have a more interesting solution which covers both subtype and bind-param constraints, so going to close this one and open separate PR for that.
